TESTING.adoc

```diff
@@ -0,0 +1,107 @@
+    Given Minishift has state "Running"
+     When executing "minishift addons apply cors" succeeds
+     Then stdout should contain "CORS is now allowed from any address"
+```
+
+[NOTE]
+====
+We recommend aligning white space to the ends of the Gherkin keywords `Given`, `When`, `Then`, and `And`.
+====
+
+See the link:https://github.com/cucumber/cucumber/wiki/Gherkin[Gherkin wiki] for more information about Gherkin.
+More in-depth information about the Minishift integration test suite can be found in the link:https://docs.openshift.org/latest/minishift/contributing/developing.html#integration-tests[Developing Minishift] documentation.
+
+[[structure-of-tests]]
+=== Structure of Tests
+Feature files which define the test cases for each individual add-on are stored in the *_test/integration/features_* directory.
+The feature files follow the naming convention of `<addon-name>.feature` and begin with the tag `@<addon-name>`.
+
+Integration tests are implemented in the *_test/integration/integration_test.go_* file.
+This file imports the test suite from the Minishift repository and can be used for the import of additional step definitions.
+
+[[prerequisites]]
+== Prerequisites
+You must do the following to run the tests successfully:
+
+- Install link:https://docs.openshift.org/latest/minishift/getting-started/installing.html[Minishift and its prerequisites].
+- Install a recent link:https://golang.org/dl/[Go distribution] (>=1.8).
+- Install the link:https://github.com/golang/dep[dep package] with the following: `go get -u github.com/golang/dep/cmd/dep`
+- `git clone` this repository into *_$GOPATH/src/github.com/minishift/minishift-addons_*.
+- Get the latest `minishift` binary from https://github.com/minishift/minishift/releases/latest and place it in *_testing/bin_*.
+- Vendor the needed Go packages by runining `make vendor` in the root directory of the *minishift-addons* repository.
+
+[NOTE]
+====
+We recommend vendoring the packages after every `git pull` from the *minishift-addons* master branch.
+====
+
+[[running-tests]]
+== Running Tests
+To run integration tests for a specific add-on, go to the root directory of the *minishift-addons* repository and run the following command: 
+
+```
+make integration ADDON=<name-of-addon>
+```
+
+This command will start tests on all Gherkin feature files present in *_tests/integration/features_* which contain the tag `@<name-of-addon>`.
+For most add-ons, this will mean using one feature file located at *_tests/integration/features/<name-of-addon>.feature_*.
+
+To run the tests on all add-ons, you can use the following:
+
+```
+make integration
+```
+
+[[additional-options]]
+=== Additional Options
+It is possible to run the tests with a Minishift binary located outside of the default location of *_testing/bin_*.
+To do so, use the `MINISHIFT_BINARY` parameter with the `make integration` target.
+For example:
+
+```
+make integration ADDON=xpaas MINISHIFT_BINARY=/home/user/minishift/minishift
```

Comment:
It is high chance that user will clone this repo in `$GOPATH/src/minishift/` path.
Let's add the following as part of the documentation
```
$ git clone https://github.com/minishift/minishift-addons $GOPATH/src/github.com/minishift/minishift-addons
```
as I was getting following errors on initial run(since I had clone it in different location):
```
$ make integration MINISHIFT_BINARY=/Users/devtools/gowork/bin/minishift
mkdir -p /Users/devtools/redhat/minishift-addons/testing/integration-test
go test -timeout 3600s github.com/minishift/minishift-addons/test/integration --tags=integration -v -args --test-dir /Users/devtools/redhat/minishift-addons/testing/integration-test --binary /Users/devtools/gowork/bin/minishift \
	--run-before-feature="" --test-with-specified-shell="" --tags=""
can't load package: package github.com/minishift/minishift-addons/test/integration: cannot find package "github.com/minishift/minishift-addons/test/integration" in any of:
	/Users/devtools/go/src/github.com/minishift/minishift-addons/test/integration (from $GOROOT)
	/Users/devtools/gowork/src/github.com/minishift/minishift-addons/test/integration (from $GOPATH)
make: *** [integration] Error 1
```

Reply:
Thanks @budhrg for noticing. Going now to implement the suggestions.

Reply:
@budhrg Actually just realized, this topic is covered in "Prerequisites" chapter:
```
- `git clone` this repository into *_$GOPATH/src/github.com/minishift/minishift-addons_*.
```